### PR TITLE
Statements before and between member initializers in special member functions

### PR DESCRIPTION
**Pull request by jbatez**
This is a proposed solution for Issue https://github.com/hsutter/cppfront/issues/475

Instead of allowing only `out` parameter assignments before and between member initializers in special member functions, this change allows _all_ statements. When generating constructors, it uses IIFE's and the comma operator to insert statements in the member initializer list. And when generating assignment operators, it does exactly what you'd expect.

Since statements can now go before or after member initializers, it leads to the question: where do statements go in a function with only implicit member initializers? I came up with the following rules:

1. By default, statements are inserted as late as possible while still maintaining the given order relative to explicit member initializers. This means a function without any explicit member initializers will have its statements inserted after all the implicit initializers.
2. To insert statements before the implicit initializers, add a special `this = that` statement after the leading statements. Note: this only works for functions that include `that` as a parameter. If we want something similiar for functions without `that`, we could add a `this = {}` or `this = default` syntax, but I have a harder time imagining real-world use-cases for that.

This change also makes implicit `this.member = that.member` initialization opt-in. Without an explicit `this = that` statement in the body, `that` functions default to the same default initialization behavior as all the other special member functions. I personally find this more consistent with the rest of the language and less surprising, but I'm open to other ideas. I've updated the `copyable` metatype accordingly.

I've also gotten rid of implicit member assignment in explicit assignment (i.e. `inout this`) operators. The way I see it, if you're writing custom assignment operators, it's because there's something special about your type that can't be described by defaults. Implicit assignment operators (i.e. the ones that get generated alongside constructors when you write `out this`) still assign to every member to match the corresponding constructor, and you can still use `this = that` to opt-in to implicit memberwise copy or move.

This change is a work-in-progress. I'm still working on cleaning up the existing regression tests and adding new tests and will likely find a few more bugs or edge cases I hadn't considered. And if you accept this approach, I need to update the documentation.

One obvious design hole is: what should we do about declarations between initializers? They can't escape the IIFE's in the cpp1 initializer lists, but they're definitely useful in assignment operators (see the example in the linked issue).

Edit: `return` statements before and between member initialzers also raise some interesting questions.

### CODE REVIEW

**Comment by jbatez** [comment]
I just realized I don't need to use the comma operator, I can just return the initializer from the lambda. That way, variables declared before a member initializer can be used in that member initializer. E.g.:

````
t: type = {
    m: i32;
    operator=: (out this, a: i32) = {
        x: i32 = a;
        m = x;
    }
}
````

generates:

```
    t::t(cpp2::in<cpp2::i16> a)
        : m{ [&]() -> cpp2::i16 { cpp2::i16 x {a}; return std::move(x); }() }
    {
    }

    auto t::operator=(cpp2::in<cpp2::i16> a) -> t& {
        cpp2::i16 x {a}; 
        m = std::move(x);
        return *this;
    }
```

**Comment by JohelEGP** [comment]
> One obvious design hole is: what should we do about declarations between initializers? They can't escape the IIFE's in the cpp1 initializer lists, but they're definitely useful in assignment operators (see the example in the linked issue).

My understanding it that permitting that just comes naturally to Cpp2.
But what Cppfront can do is limited, given that it has to lower to Cpp1.

What if you delegated to a generated constructor to tap into the power of the stack?
Have this:
```Cpp2
t: type = {
  public x: i32;
  operator=: (out this, y: i32) = {
    z := y;
    x = z;
  }
}
```
Generate this:
```C++
class t {
  public: cpp2::i32 x;
  struct cpp2_stack_3_2 { };
  public: explicit t(cpp2::in<cpp2::i32> y);
  private: explicit t(cpp2::in<cpp2::i32> y, cpp2_stack_3_2, auto);
  public: auto operator=(cpp2::in<cpp2::i32> y) -> t& ;

  public: t(t const&) = delete; /* No 'that' constructor, suppress copy */
  public: auto operator=(t const&) -> void = delete;
};

  t::t(cpp2::in<cpp2::i32> y)
                                    : t{ y, cpp2_stack_3_2{}, [] {
    struct cpp2_stack_t {
      cpp2::deferred_init<cpp2::i32> z;
    };
    return cpp2_stack_t{};
                                    }() }
  {}
  t::t(cpp2::in<cpp2::i32> y, t::cpp2_stack_3_2, auto cpp2_stack)
                                    : x{ (cpp2_stack.z.construct(y), cpp2_stack.z.value()) }
  {}
  auto t::operator=(cpp2::in<cpp2::i32> y) -> t& {
                                    x = y;
                                    return *this;
  }
```
See <https://cpp2.godbolt.org/z/EGb3jY8zr>.

**Comment by jbatez** [comment]
That's a clever trick. It gets complicated pretty quickly, though; you have to track down all the references to `z` and replace with `cpp2_stack.z.value()`. E.g.

```
t: type = {
  public x: i32;
  operator=: (out this, y: i32) = {
    z := y;
    std::cout << z << std::endl;
    w := z;
    x = w;
  }
}
```

Generate:

```
class t {
  public: cpp2::i32 x;
  private: struct cpp2_stack_3_2 { };
  public: explicit t(cpp2::in<cpp2::i32> y);
  private: explicit t(cpp2::in<cpp2::i32> y, cpp2_stack_3_2, auto);
  public: auto operator=(cpp2::in<cpp2::i32> y) -> t& ;

  public: t(t const&) = delete; /* No 'that' constructor, suppress copy */
  public: auto operator=(t const&) -> void = delete;
};

  t::t(cpp2::in<cpp2::i32> y)
                                    : t{ y, cpp2_stack_3_2{}, [] {
    struct cpp2_stack_t {
      cpp2::deferred_init<decltype(auto{y})> z;
      cpp2::deferred_init<decltype(auto{z.value()})> w;
    };
    return cpp2_stack_t{};
                                    }() }
  {}
  t::t(cpp2::in<cpp2::i32> y, t::cpp2_stack_3_2, auto cpp2_stack)
                                    : x{ [&]() -> cpp2::i32 {
                                          cpp2_stack.z.construct(y);
                                          std::cout << cpp2_stack.z.value() << std::endl;
                                          cpp2_stack.w.construct(cpp2_stack.z.value());
                                          return cpp2_stack.w.value();
                                      }() }
  {}
  auto t::operator=(cpp2::in<cpp2::i32> y) -> t& {
                                    auto z{y};
                                    x = z;
                                    return *this;
  }
```
See https://cpp2.godbolt.org/z/89EsnhMxM.

I'll give it a try, but it might be more work than I bargained for at this time.

File: source/cppfront.cpp [hunk 1]
```diff
@@ -4537,67 +4548,111 @@ class cppfront
...
+                                    stmt_pos,
+                                    "'this = that' may only appear once"
+                                );
+                                return;
+                            }
+                            else
+                            {
+                                // All the statements before 'this = that' need to be inserted
+                                // before the first member initializer whether or not it's explicit
+                                pending_statements = deferred_statements;
```
Comment (gregmarr) [inline source/cppfront.cpp:4585]: This is still a good optimization to prevent allocation:

    pending_statements = std::move(deferred_statements);

You can also do it like this:

    std::swap(pending_statements, deferred_statements);
    deferred_statements.clear();

File: source/cppfront.cpp [hunk 2]
```diff
@@ -4633,11 +4633,9 @@ class cppfront
                 // If explicit, upgrade deferred statements to pending
                 if (found_explicit_init)
                 {
-                    pending_statements.insert(
-                        pending_statements.end(),
-                        deferred_statements.begin(),
-                        deferred_statements.end()
-                    );
```
Comment (gregmarr) [inline source/cppfront.cpp]: You can make these moves and keep the `insert`.

                    pending_statements.insert(
                        pending_statements.end(),
                        std::make_move_iterator(deferred_statements.begin()),
                        std::make_move_iterator(deferred_statements.end());

**Comment by jbatez** [comment]
I think I'm done with this for now. I didn't do the work to handle things declared before a member initializer and used after. The more I thought about it, the more complicated it got. I think it'd be awesome if the feature was there, but I unfortunately don't have the time to implement it myself.

File: source/cppfront.cpp [hunk 3]
```diff
@@ -4684,40 +4752,28 @@ class cppfront
...
-                        auto ret = std::string{};
-                        for (auto& init : out_inits) {
-                            ret += init + ", ";
+                    //  If there are any pending statements, wedge them into
+                    //  this initializer using an immediately invoked lambda
+                    auto& mem_inits = current_functions.back().prolog.mem_inits;
+                    if (!pending_statements.empty())
+                    {
+                        auto object_type = print_to_string( *(*object)->get_object_type() );
+                        mem_inits.push_back(separator + object_name + "{ [&]() -> " + object_type + " {");
```
Comment (JohelEGP) [inline source/cppfront.cpp:4761]: Fixing the return type to the type of the data member to initialize is NG.
This should achieve the semantics one would expected reading the Cpp2 code:
```suggestion
                        mem_inits.push_back(separator + object_name + "{ [&]() -> decltype(auto) {");
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

**Comment by jbatez** [comment]
The original issue has been addressed by https://github.com/hsutter/cppfront/commit/cdf71bdca64536a005f2491d8c19f1d05a1c62f6

**Comment by hsutter** [comment]
Thanks, Jo!
